### PR TITLE
Allow date and term calculation in liquid

### PR DESCRIPTION
**Pull request by sumanth-spotdraft**
(A) If date = 31 March 2020 then 30 day period = Print date + 30 days . 
(B) If Date 1 = 31 March 2020 and Date 2 = 30 March 2021 then Print Duration between Date 1 and Date 2

### CODE REVIEW

**Comment by codebreach** [comment]
Tests? Also add date but subtract something else. please

File: test/filters.js [hunk 1]
```diff
@@ -205,6 +186,13 @@ describe('filters', function () {
       () => test('{{ 183.357 | minus: 12 }}', '171.357'))
     it('should convert first arg as number', () => test('{{ "4" | minus: 1 }}', '3'))
     it('should convert both args as number', () => test('{{ "4" | minus: "1" }}', '3'))
+    it('should return 365 days', function () {
+      var fromDate = new Date("March 31, 2020");
+      var toDate = new Date("March 31, 2021");
+      var src = '{{' + fromDate + '| minus: ' + toDate + '}}';
+      var dst = 365
```
Comment (codebreach) [inline test/filters.js]: this should return a duration not just a plain number

File: filters.js [hunk 2]
```diff
@@ -101,8 +101,8 @@ var filters = {
   },
   upcase: str => stringify(str).toUpperCase(),
   url_encode: encodeURIComponent,
-  addDuration: (v, arg) => addDateDuration(v, arg),
-  subtractDates: (v, arg) => subtractDateDuration(v, arg)
+  add_duration: (v, arg) => addDateDuration(v, arg),
+  duration_in_days: (v, arg) => durationBetweenDates(v, arg)
```
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply filters.js]: Overloading the existing plus and minus decorators

File: filters.js [hunk 3]
```diff
@@ -101,8 +101,8 @@ var filters = {
   },
   upcase: str => stringify(str).toUpperCase(),
   url_encode: encodeURIComponent,
-  addDuration: (v, arg) => addDateDuration(v, arg),
-  subtractDates: (v, arg) => subtractDateDuration(v, arg)
+  add_duration: (v, arg) => addDateDuration(v, arg),
```
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply filters.js]: Overloading the existing plus and minus decorators

File: test/filters.js [hunk 4]
```diff
@@ -116,6 +116,27 @@ describe('filters', function () {
...
+      return test('{{ new Date() | add_duration: {value: 1, type: "days"} }}')
+    })
+    it('should add 1 year to current date', () => {
+      return test('{{ new Date() | add_duration: {value: 1, type: "years"} }}')
+    })
+  })
+
+  describe('duration_in_days', () => {
+    it('should return 365 days', () => {
+      return test('{{new Date("31 March 2020") | duration_in_days: new Date("31 March 2021")}}')
```
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply test/filters.js]: updated

File: filters.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply filters.js]: Overloading the existing plus and minus decorators

File: filters.js [hunk 5]
```diff
@@ -167,8 +173,34 @@ function add(v, arg) {
...
+
+function durationBetweenDates(v, arg) {
+  return performOperations(v, arg, "DURATION_IN_DAYS");
+}
+
 function performOperations(v, arg, operation) {
-  if (typeof v === "object" && typeof arg === "object") {
+  /**
+   * Check Date and add period.
+   * {% date | add_duration : {value: 10, type: "week"} %}
```
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply filters.js]: Overloading the existing plus and minus decorators

File: test/filters.js (hunk 1 above)
Comment (sumanth-spotdraft) [reply test/filters.js]: Resolved

File: filters.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply filters.js]: removed this and using the existing decorators.

File: filters.js [hunk 6]
```diff
@@ -167,8 +173,34 @@ function add(v, arg) {
...
+    }
+  }
+  /**
+   * Check Date and calculate duration between dates.
+   * {% date1 | DURATION_IN_DAYS: date2 %}
+   */
+  else if(Object.prototype.toString.call(v) === '[object Date]' && Object.prototype.toString.call(arg) === '[object Date]') {
+    return operationOnItem(v, arg, operation);
+  }
+  else if (typeof v === "object" && typeof arg === "object") {
```
[the remark this answers is not in the document]
Comment (sumanth-spotdraft) [reply filters.js]: Added support for adding duration.

File: filters.js [hunk 7]
```diff
@@ -167,8 +194,24 @@ function add(v, arg) {
...
 function performOperations(v, arg, operation) {
+  if(Object.prototype.toString.call(v) === '[object Date]' && Object.prototype.toString.call(arg) === '[object Date]') {
+    return operationOnDate(v, arg, operation);
+  }
+  if (Object.prototype.toString.call(v) === '[object Date]' && typeof arg === "object") {
+    const addType = ['days', 'weeks', 'months', 'years'];
+    const {value, type} = arg;
+    if (value && addType.indexOf(type) != -1) {
+      return operationOnDate(v, arg, operation);
+    }
```
Comment (codebreach) [inline filters.js]: else raise error

**Comment by sumanth-spotdraft** [comment]
> Tests? Also add date but subtract something else. please

Added

File: test/filters.js [hunk 8]
```diff
@@ -184,6 +194,14 @@ describe('filters', function () {
       () => test('{{ 183.357 | minus: 12 }}', '171.357'))
     it('should convert first arg as number', () => test('{{ "4" | minus: 1 }}', '3'))
     it('should convert both args as number', () => test('{{ "4" | minus: "1" }}', '3'))
+    it('should return {"type":"days","value":730}', () => {
+      try {
+        const dst = {type: "days", value: 730};
+        return test('{{ from_date | minus: to_date}}', JSON.stringify(dst))
```
Comment (codebreach) [inline test/filters.js]: since we expect this to be used with `assign` it will be nice if we can add a test for that?

**Pull request merged**